ruby/phase-1.md

```diff
@@ -0,0 +1,536 @@
+  namespace: 'my-namespace',
+  task_queue: 'my-queue'
+})
+
+worker = Temporal::Worker.new(config, 'my-namespace', 'my-queue')
+worker.register_workflow(MyWorkflow)
+worker.register_activity(MyActivity)
+worker.run
+```
+
+
+## Client
+
+The client is the primary point for interacting with workflows and async activities as well as
+managing namespaces. A client is created with a configuration object that specifies options shared
+between a worker and a client.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7233')
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow, 'arg_1', kwarg_2: 'foo', options: {
+  namespace: 'my-namespace',
+  task_queue: 'my-queue'
+})
+```
+
+Notes:
+
+- Client is intended to be multi-tenant capable (no limit on a single namespace)
+- Configuration will provide defaults where possible (avoiding extensive upfront configuration)
+- Configuration will be used for other more advanced options such as logger, metrics, payload
+  converters, middleware, etc
+
+```ruby
+class Temporal::Client
+  # Creates a new instance of a Client
+  def self.new(config: Temporal::Configuration, **options) -> Temporal::Client
+
+  # Start the workflow and return handle for further interaction
+  # TODO: expand on :options
+  def start_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    options: {}, # execution options
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> Temporal::WorkflowHandle
+
+  # Start the workflow and block awaiting it's completion
+  def execute_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    options: {}, # execution options
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> Any
+
+  # Generate a handle object from a namespace, workflow_id and a run_id
+  def workflow_handle(namespace, workflow_id, run_id = nil) -> Temporal::WorkflowHandle
+end
+```
+
+Notes:
+
+- This is not a complete definition of the client
+- Both positional arguments and keyword arguments are supported, except for the reserved `options:`
+- Conventions similar to Python SDK are used here:
+  - `Client` is a general purpose interface, not limited workflow-only interactions
+  - Due to this method names are more descriptive than just `#start`, `#execute`, etc
+  - `Client#execute_workflow` is added for convenience
+
+Once a workflow is started all the interactions will be going through the handle object returned to
+the caller. The handle can of course be created from a workflow_id/run_id combination using the
+`Client#workflow_handle` method. Here the interface of the handle:
+
+```ruby
+class Temporal::WorkflowHandle
+  # Return the result of the workflow execution or raise if workflow is still running
+  # - await: true will wait (for up to :timeout seconds) for the workflow result
+  def result(await: false, timeout: nil) -> Any
```

Comment:
Good point on making `await: true` by default, I'll make the change.
In terms of timeout — it will get translated into a gRPC deadline, but why do you think it shouldn't be a part of this interface?

I'll check what `follow_runs` means, I haven't seen that option